File: tools/showcase_state_motion.py
"""Derived, inspectable endpoint geometry for the ABL showcase; no trajectory claim."""
from __future__ import annotations

import hashlib
import json
import math
from pathlib import Path

import numpy as np
from Bio.PDB import MMCIFParser
from Bio.PDB.vectors import calc_angle, calc_dihedral


def fit(moving, fixed):
    """Proper row-vector Kabsch transform, with reflections forbidden."""
    mc, fc = moving.mean(axis=0), fixed.mean(axis=0)
    u, _, vt = np.linalg.svd((moving - mc).T @ (fixed - fc))
    if np.linalg.det(u @ vt) < 0:
        u[:, -1] *= -1
    rotation = u @ vt
    return rotation, fc - mc @ rotation


def angles(residues, position):
    r = residues[position]
    out = {"phi": None, "psi": None, "chi1": None, "n_ca_c": None}
    def measure(names, key, angle=False):
        try:
            atoms = [residues[p][a] for p, a in names]
            if any(a.is_disordered() for a in atoms):
                return
            value = (calc_angle if angle else calc_dihedral)(*[a.get_vector() for a in atoms])
            if math.isfinite(value):
                out[key] = round(math.degrees(value), 2)
        except KeyError:
            pass
    # Never calculate a torsion across an absent residue or broken peptide bond.
    if position - 1 in residues and "C" in residues[position - 1] and "N" in r:
        if 1.0 < np.linalg.norm(residues[position - 1]["C"].coord - r["N"].coord) < 1.9:
            measure([(position - 1, "C"), (position, "N"), (position, "CA"), (position, "C")], "phi")
    if position + 1 in residues and "C" in r and "N" in residues[position + 1]:
        if 1.0 < np.linalg.norm(r["C"].coord - residues[position + 1]["N"].coord) < 1.9:
            measure([(position, "N"), (position, "CA"), (position, "C"), (position + 1, "N")], "psi")
    # Restrict chi1 to DFG residues with the unambiguous N-CA-CB-CG definition.
    if r.resname in {"ASP", "PHE"}:
        measure([(position, a) for a in ["N", "CA", "CB", "CG"]], "chi1")
    measure([(position, a) for a in ["N", "CA", "C"]], "n_ca_c", angle=True)
    return out


def pdb_text(residues, rotation, translation):
    lines, serial, previous = [], 0, None
    for pos, res in sorted(residues.items()):
        if previous is not None and pos != previous + 1:
            lines.append("TER")
        for atom in res:
            if atom.is_disordered():
                continue
            serial += 1
            x, y, z = atom.coord.astype(float) @ rotation + translation
            lines.append(f"ATOM  {serial:5d} {atom.name:>4} {res.resname:>3} A{pos:4d}    "
                         f"{x:8.3f}{y:8.3f}{z:8.3f}{1.:6.2f}{atom.bfactor:6.2f}          {atom.element:>2}")
        previous = pos
    return "\n".join(lines + ["TER", "END", ""])


def build_motion(panel: Path):
    structures, provenance, exclusions = [], [], []
    for entry, state in [("2GQG", "active"), ("2HYY", "inactive")]:
        source = panel / f"sources/wwpdb/{entry}.cif"
        mapping = panel / f"references/abl/alignment_map.{entry}.json"
        run = panel / f"results/execution-abl/v2-stateatlas-{state}-held_out-{entry}/RUN_MANIFEST.json"
        hashes = json.loads(run.read_text())["input_sha256"]
        for path, key in [(source, "structure"), (mapping, "alignment_map")]:
            digest = hashlib.sha256(path.read_bytes()).hexdigest()
            if digest != hashes[key]:
                raise ValueError(f"State motion input drift: {path.relative_to(panel)}")
            provenance.append({"path": str(path.relative_to(panel)), "sha256": digest})
        doc = json.loads(mapping.read_text())
        if doc["coordinate_system"] != "uniprot" or doc["domain"] != {"uniprot_start": 242, "uniprot_end": 495}:
            raise ValueError("Unexpected state motion coordinate system or domain")
        chain = MMCIFParser(QUIET=True).get_structure(entry, source)[0]["A"]
        mapped = {}
        for row in doc["query"]:
            pos = row["uniprot_position"]
            if row["mapping_state"] != "exact" or row.get("chain_id") != "A":
                continue
            key = (" ", int(row["auth_seq_id"]), row.get("insertion_code") or " ")
            if key in chain and "CA" in chain[key] and not chain[key]["CA"].is_disordered():
                mapped[pos] = chain[key]
        structures.append(mapped)
    a, b = structures
    positions = sorted(p for p in a.keys() & b.keys() if a[p].resname == b[p].resname)
    exclusions = sorted(set(range(242, 496)) - set(positions))
    if len(positions) / 254 < .9:
        raise ValueError("State motion shared exact mapping covers less than 90% of domain")
    a = {p: a[p] for p in positions}; b = {p: b[p] for p in positions}
    active = np.array([a[p]["CA"].coord for p in positions], dtype=float)
    inactive = np.array([b[p]["CA"].coord for p in positions], dtype=float)
    rotation, translation = fit(inactive, active)
    placed = inactive @ rotation + translation
    displacement = np.linalg.norm(active - placed, axis=1)
    residues = []
    for i, pos in enumerate(positions):
        residues.append({"position": pos, "resname": a[pos].resname,
                         "active": active[i].round(5).tolist(), "inactive": placed[i].round(5).tolist(),
                         "displacement": round(float(displacement[i]), 4),
                         "active_angles": angles(a, pos), "inactive_angles": angles(b, pos)})
    regions = [{"id": "activation", "label": "Activation loop / DFG · 381–402", "start": 381, "end": 402},
               {"id": "dfg", "label": "DFG motif · Asp381–Phe382–Gly383", "start": 381, "end": 383},
               {"id": "alpha", "label": "αC region · 280–299", "start": 280, "end": 299},
               {"id": "ploop", "label": "P-loop region · 248–255", "start": 248, "end": 255},
               {"id": "all", "label": "Whole mapped kinase domain · 242–495", "start": 242, "end": 495}]
    for region in regions:
        selected = [r["displacement"] for r in residues if region["start"] <= r["position"] <= region["end"]]
        region["mean_displacement"] = round(float(np.mean(selected)), 4)
        region["mapped_count"] = len(selected)
    return {"schema_version": 1, "active_entry": "2GQG", "inactive_entry": "2HYY", "chain": "A",
            "coordinate_system": "UniProt P00519; exact frozen query maps; first coordinate model",
            "method": "Unweighted CA Kabsch fit of inactive onto active over all shared exact same-residue positions within 242–495; no flexible-region exclusion or outlier rejection.",
            "claim": "New derived endpoint geometry for teaching; not a StateAtlas classification rerun, molecular-dynamics trajectory or transition pathway.",
            "source_inputs": provenance, "mapped_count": len(positions), "excluded_positions": exclusions,
            "exclusion_note": "Missing, nonstandard, ambiguous or sequence-different positions are omitted. Tyr393 is phosphorylated (PTR) in 2GQG and is not bridged by the motion trace. Disordered atoms are excluded from angle measurements and exported models.",
            "rmsd": round(float(np.sqrt(np.mean(displacement ** 2))), 4),
            "rotation": rotation.tolist(), "translation": translation.tolist(),
            "regions": regions, "residues": residues,
            "active_pdb": pdb_text(a, np.eye(3), np.zeros(3)),
            "inactive_pdb": pdb_text(b, rotation, translation)}

File: tools/showcase_state_motion.js
// Endpoint interpolation is illustrative; it is never a physical trajectory.
let stopStateMotion = () => {};
function buildStateMotion(p) {
  const m=p.motion, el=document.getElementById('state-motion-view');
  const viewer=$3Dmol.createViewer(el,{backgroundColor:bg(),antialias:true});
  viewer.addModel(m.active_pdb,'pdb'); viewer.addModel(m.inactive_pdb,'pdb');
  const slider=document.getElementById('state-motion-slider');
  const region=document.getElementById('state-motion-region');
  const residue=document.getElementById('state-motion-residue');
  const play=document.getElementById('state-motion-play');
  const status=document.getElementById('state-motion-status');
  let timer=null,direction=1;
  const vec=a=>({x:a[0],y:a[1],z:a[2]});
  const fmt=n=>n===null?'unavailable':n.toFixed(1)+'°';
  const difference=(a,b)=>a===null||b===null?'unavailable':(((b-a+540)%360)-180).toFixed(1)+'°';
  function selection(){const r=m.regions.find(x=>x.id===region.value);return m.residues.filter(x=>x.position>=r.start&&x.position<=r.end);}
  function detail(){
    const r=m.residues.find(x=>x.position===Number(residue.value));
    document.getElementById('state-motion-detail').textContent=
      r.resname+' '+r.position+': Cα displacement '+r.displacement.toFixed(2)+' Å after the declared fit.';
    const body=document.getElementById('state-angle-body');body.replaceChildren();
    for(const [key,label] of [['phi','Backbone φ (torsion)'],['psi','Backbone ψ (torsion)'],['chi1','Side-chain χ1 (Asp/Phe only)'],['n_ca_c','N–Cα–C bond angle']]){
      const a=r.active_angles[key],b=r.inactive_angles[key],tr=document.createElement('tr');
      for(const value of [label,fmt(a),fmt(b),key==='n_ca_c'?(a===null||b===null?'unavailable':(b-a).toFixed(1)+'°'):difference(a,b)]){
        const td=document.createElement('td');td.textContent=value;tr.appendChild(td);
      }body.appendChild(tr);
    }
  }
  function draw(){
    const t=Number(slider.value)/100, selected=selection(), ids=selected.map(r=>r.position);
    viewer.removeAllShapes(); viewer.removeAllLabels();
    viewer.setStyle({},{cartoon:{color:'#9CA3AF',opacity:0.16}});
    viewer.setStyle({model:0,resi:ids},{cartoon:{color:'#2A9D8F',opacity:0.4}});
    viewer.setStyle({model:1,resi:ids},{cartoon:{color:'#7B5EA7',opacity:0.4}});
    const focus=Number(residue.value);
    viewer.addStyle({model:0,resi:focus},{stick:{color:'#2A9D8F',radius:0.18}});
    viewer.addStyle({model:1,resi:focus},{stick:{color:'#7B5EA7',radius:0.18}});
    let previous=null;
    for(const r of selected){
      const xyz=r.active.map((x,i)=>x*(1-t)+r.inactive[i]*t),color=r.position===focus?'#E63946':'#E9A23B';
      viewer.addSphere({center:vec(xyz),radius:r.position===focus?0.65:0.28,color,clickable:true,
        callback:()=>{residue.value=String(r.position);detail();draw();}});
      if(previous&&r.position===previous.r.position+1&&
        Math.hypot(...r.active.map((x,i)=>x-previous.r.active[i]))<4.5&&
        Math.hypot(...r.inactive.map((x,i)=>x-previous.r.inactive[i]))<4.5){
        viewer.addCylinder({start:vec(previous.xyz),end:vec(xyz),radius:0.12,color:'#E9A23B',fromCap:1,toCap:1});
      }
      if(r.position===focus){
        if(r.displacement>0.1)viewer.addArrow({start:vec(r.active),end:vec(r.inactive),radius:0.13,color:'#E63946'});
        viewer.addLabel(r.resname+' '+r.position,{position:vec(xyz),fontSize:12,fontColor:'#111',backgroundColor:'#fff',backgroundOpacity:0.9,inFront:true});
      }
      previous={r,xyz};
    }
    status.textContent=t===0?'Active-like endpoint · 2GQG':t===1?'Inactive-like endpoint · 2HYY':
      'Illustrative interpolation · '+slider.value+'% · not an observed structure';
    viewer.render();
  }
  function reset(){
    const selected=selection();
    if(!selected.some(r=>r.position===Number(residue.value)))residue.value=String(selected.reduce((a,b)=>a.displacement>b.displacement?a:b).position);
    draw();viewer.zoomTo({resi:selected.map(r=>r.position)});viewer.zoom(region.value==='dfg'?1.15:0.72);viewer.render();detail();
    const r=m.regions.find(x=>x.id===region.value);
    document.getElementById('state-region-detail').textContent=r.mapped_count+' mapped residues; mean endpoint displacement '+r.mean_displacement.toFixed(2)+' Å. Gaps are left open.';
  }
  stopStateMotion=()=>{if(timer)clearInterval(timer);timer=null;play.textContent='Play illustrative motion';play.setAttribute('aria-pressed','false');};
  play.addEventListener('click',()=>{if(timer){stopStateMotion();return;}play.textContent='Pause';play.setAttribute('aria-pressed','true');
    timer=setInterval(()=>{let n=Number(slider.value)+direction*2;if(n>=100){n=100;direction=-1;}if(n<=0){n=0;direction=1;}slider.value=String(n);draw();},90);});
  slider.addEventListener('input',()=>{stopStateMotion();draw();});
  region.addEventListener('change',()=>{stopStateMotion();reset();});
  residue.addEventListener('change',()=>{const n=Number(residue.value),r=m.regions.find(x=>x.id===region.value);
    if(n<r.start||n>r.end)region.value='all';reset();});
  document.getElementById('state-motion-reset').addEventListener('click',reset);
  document.getElementById('state-motion-export').addEventListener('click',()=>{
    const url=URL.createObjectURL(new Blob([JSON.stringify(m,null,2)],{type:'application/json'}));
    const a=document.createElement('a');a.href=url;a.download='ABL-endpoint-geometry.json';a.click();setTimeout(()=>URL.revokeObjectURL(url),1000);
  });
  document.addEventListener('visibilitychange',()=>{if(document.hidden)stopStateMotion();});
  reset();made[p.id].push(viewer);
}
